app/src/test/kotlin/compiler/lexer/lexer_grammar/RegexParserTest.kt

```diff
@@ -0,0 +1,114 @@
+package compiler.lexer.lexer_grammar
+
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertFails
+
+internal class RegexParserTest {
+
+    class TestRegexParser : UniversalRegexParser<String>() {
+        public override fun performStar(child: String): String {
+            return "(*$child*)"
+        }
+
+        public override fun performConcat(left: String, right: String): String {
+            return "(?$left,$right?)"
+        }
+
+        public override fun performUnion(left: String, right: String): String {
+            return "(|$left,$right|)"
+        }
+
+        public override fun getEmpty(): String {
+            return "EMP"
+        }
+
+        public override fun getAtomic(charSet: Set<Char>): String {
+            return if (charSet.size == 1) charSet.toList()[0].toString()
+            else "(&" + charSet.toList().sorted().joinToString("") + "&)"
+        }
+    }
+
+    companion object {
+        val PARSER = TestRegexParser()
+    }
+
+    @Test
+    fun `test basic operations`() {
+        val star = PARSER.parseStringToRegex("a*")
+        val concat = PARSER.parseStringToRegex("ab")
+        val union = PARSER.parseStringToRegex("a|b")
+        val brackets = PARSER.parseStringToRegex("[abc]")
+        var bracketsExpected = PARSER.getEmpty()
+        for (char in "abc") bracketsExpected = PARSER.performUnion(bracketsExpected, char.toString())
+
+        assertEquals("(*a*)", star)
+        assertEquals("(?a,b?)", concat)
+        assertEquals("(|a,b|)", union)
+        assertEquals(bracketsExpected, brackets)
+    }
+
+    @Test
+    fun `test concatenation`() {
+        val expressionsToConcatenate = listOf("a", "b*", "(x|d)", "[dd]", "(d)*", "(xd)")
+        val expressionsParsed = expressionsToConcatenate.map { PARSER.parseStringToRegex(it) }
+
+        val concatenationParsed = PARSER.parseStringToRegex(expressionsToConcatenate.joinToString(""))
+        var expectedResult = expressionsParsed[0]
+        for (i in 1 until expressionsParsed.size) expectedResult = PARSER.performConcat(expectedResult, expressionsParsed[i])
```

Comment:
Can you use `reduce` syntax? Like `val expectedResult = expressionsParsed.reduce { acc, next -> PARSER.performConcat(acc, next) }
`